src/logger/logger.cpp

```diff
@@ -0,0 +1,84 @@
+#include "logger.hh"
+
+#include <cstdarg>
+#include <iomanip>
+#include <iostream>
+#include <filesystem>
+#include <string>
+#include <thread>
+
+ZarrLogLevel Logger::current_level_ = ZarrLogLevel_Info;
+std::mutex Logger::log_mutex_{};
+
+void
+Logger::set_log_level(ZarrLogLevel level)
+{
+    current_level_ = level;
+}
+
+ZarrLogLevel
+Logger::get_log_level()
+{
+    return current_level_;
+}
+
+std::string
+Logger::log(ZarrLogLevel level,
+            const char* file,
+            int line,
+            const char* func,
+            const char* format,
+            ...)
+{
+    std::scoped_lock lock(log_mutex_);
+    if (current_level_ == ZarrLogLevel_None || level < current_level_) {
+        return {}; // Suppress logs
+    }
+
+    va_list args;
+    va_start(args, format);
+
+    std::string prefix;
+    std::ostream* stream = &std::cout;
+
+    switch (level) {
+        case ZarrLogLevel_Debug:
+            prefix = "[DEBUG] ";
+            break;
+        case ZarrLogLevel_Info:
+            prefix = "[INFO] ";
+            break;
+        case ZarrLogLevel_Warning:
+            prefix = "[WARNING] ";
+            stream = &std::cerr;
+            break;
+        case ZarrLogLevel_Error:
+            prefix = "[ERROR] ";
+            stream = &std::cerr;
+            break;
```

Comment:
Add `default` case for completeness. Are we using a code linter like clangd/clang-tidy or sonar? This issue should be flagged by such tools.